auth/oidc.go

```diff
@@ -212,73 +238,148 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
+	return fmt.Sprintf("%s_%s", provider.UserPrefix, url.QueryEscape(subject))
+}
 
+// Clone of provider metadata construction from go-oidc library.
+// https://github.com/coreos/go-oidc/blob/v2/oidc.go#L113
+func (op *OIDCProvider) FetchCustomProviderConfig(ctx context.Context, discoveryURL string) (*oidc.Provider, error) {
 	// If discovery URL is empty, use the standard discovery URL
 	if discoveryURL == "" {
 		discoveryURL = strings.TrimSuffix(op.Issuer, "/") + discoveryConfigPath
 	}
 
 	base.Debugf(base.KeyAuth, "Fetching custom provider config from %s", base.UD(discoveryURL))
-	req, err := http.NewRequest("GET", discoveryURL, nil)
+	req, err := http.NewRequest(http.MethodGet, discoveryURL, nil)
 	if err != nil {
 		base.Debugf(base.KeyAuth, "Error building new request for URL %s: %v", base.UD(discoveryURL), err)
 		return nil, err
 	}
-	resp, err := http.DefaultClient.Do(req)
+	resp, err := doRequest(ctx, req)
 	if err != nil {
 		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %v", base.UD(discoveryURL), err)
 		return nil, err
 	}
-	defer func() { _ = resp.Body.Close() }()
-	if err := base.JSONDecoder(resp.Body).Decode(&customConfig); err != nil {
-		base.Debugf(base.KeyAuth, "Error parsing body %s: %v", base.UD(discoveryURL), err)
-		return nil, err
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			base.Debugf(base.KeyAuth, "Error closing response body %s: %v", base.UD(discoveryURL), err)
+		}
+	}()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error reading response body %s: %v", base.UD(discoveryURL), err)
+		return nil, fmt.Errorf("unable to read response body: %v", err)
 	}
 
-	var oidcConfig oidc.ProviderConfig
-	oidcConfig, err = customConfig.AsProviderConfig()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	var p ProviderMetadata
+	err = unmarshalResp(resp, body, &p)
 	if err != nil {
-		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %+v", base.UD(discoveryURL), err)
-		return nil, err
+		base.Debugf(base.KeyAuth, "Error parsing body %s: %v", base.UD(discoveryURL), err)
+		return nil, fmt.Errorf("oidc: failed to decode provider discovery object: %v", err)
 	}
 
-	// Set expiry on config, if defined in response header
-	if ttl, ok, _ := phttp.Cacheable(resp.Header); ok {
-		oidcConfig.ExpiresAt = time.Now().UTC().Add(ttl)
+	if p.Issuer != op.Issuer {
+		return nil, fmt.Errorf("oidc: issuer did not match the issuer returned by provider, expected %q got %q", op.Issuer, p.Issuer)
+	}
+	var algs []string
+	for _, a := range p.IdTokenSigningAlgValuesSupported {
+		if supportedAlgorithms[a] {
+			algs = append(algs, a)
+		}
 	}
 
-	base.Debugf(base.KeyAuth, "Returning config: %v", base.UD(oidcConfig))
-	return &oidcConfig, nil
+	provider := &oidc.Provider{}
+	SetUnexportedField(reflect.ValueOf(provider).Elem().FieldByName("issuer"), p.Issuer)
```

Comment:
Using reflection to set unexported fields is a red flag, as it defeats the purpose of using a library (if we're modifying unexported fields, we can no longer be sure that our usage of the library is supported/valid).

In this case, 'provider' is really just the provider config.  We're not using it outside of the client init, so I think we can just remove as a property on OIDCClient.

We still need the ability to create the idTokenVerifier, but it looks like that could be done using the (public) NewVerifier (cloning some of the functionality in provider.Verifier.